src/CompilingMatcher.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of composer/semver.
+ *
+ * (c) Composer <https://github.com/composer>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace Composer\Semver;
+
+use Composer\Semver\Constraint\CompilableConstraintInterface;
+use Composer\Semver\Constraint\Constraint;
+use Composer\Semver\Constraint\ConstraintInterface;
+use Composer\Semver\Constraint\NotCompilableConstraintException;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompilingMatcher
+{
+    private static $compiledCheckerCache = array();
+    private static $enabled = null;
+
+    private static $transOpInt = array(
+        Constraint::OP_EQ => '==',
+        Constraint::OP_LT => '<',
+        Constraint::OP_LE => '<=',
+        Constraint::OP_GT => '>',
+        Constraint::OP_GE => '>=',
+        Constraint::OP_NE => '!=',
+    );
+
+    /**
+     * Evaluates the expression: $constraint match $operator $version
+     *
+     * @param ConstraintInterface $constraint
+     * @param int                 $operator
+     * @param string              $version
+     *
+     * @return mixed
+     */
+    public static function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        if (self::$enabled === null) {
+            self::$enabled = !in_array('eval', explode(',', ini_get('disable_functions')));
+        }
+        if (!self::$enabled || !$constraint instanceof CompilableConstraintInterface) {
```

Comment:
Would it speedup things when doing this instanceof check before calling into CompilingMatcher->match() ?

Reply:
Well, built-in constraint are all compilable. So not calling the CompilingMatcher would be an edge case when a plugin implements the interface by itself.